src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -317,6 +328,20 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
         None
     }
 
+    override protected def contains0(key: A, hash: Int, level: Int): Boolean = {
+      // Note: this code is duplicated with `get0`
+      val index = (hash >>> level) & 0x1f
+      val mask = (1 << index)
+      if (bitmap == - 1) {
+        elems(index & 0x1f).contains0(key, hash, level + 5)
+      } else if ((bitmap & mask) != 0) {
+        val offset = Integer.bitCount(bitmap & (mask-1))
+        elems(offset).contains0(key, hash, level + 5)
```

Comment:
It's bound by the height of the tree.  The unrolling will make the code *very* ugly, and might even be a net loss by expanding the bytecode a lot (which can make the JIT compiler try less hard for a good optimization).  If you've still got your own branch around it might be worth a try, but I'd really try to avoid doing this.  Manually making it tail-recursive might be worth a try too.  (Loop as long as isInstanceOf gives the right answer.)